<a href="https://colab.research.google.com/github/asifshigri12/NLP/blob/main/Spam_text_case_srudy_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [32]:
# Import necessary libraries
import pandas as pd
import string
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer, WordNetLemmatizer
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report, confusion_matrix
from sklearn.preprocessing import LabelEncoder
import nltk
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [35]:
df = pd.read_csv('/content/SPAM text message 20170820 - Data.csv')
df.head()

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

##Data Preprocessing
Normalize the text data (lowercasing and punctuation removal)

In [36]:
# Normalize text data
def normalize_text(text):
    text = text.lower()  # Lowercasing
    text = text.translate(str.maketrans('', '', string.punctuation))  # Removing punctuation
    return text

df['Message'] = df['Message'].apply(normalize_text)


In [37]:
# Apply normalization to the message column
df['Message'] = df['Message'].apply(normalize_text)
df.head()

Category                                            Message
0      ham  go until jurong point crazy available only in ...
1      ham                            ok lar joking wif u oni
2     spam  free entry in 2 a wkly comp to win fa cup fina...
3      ham        u dun say so early hor u c already then say
4      ham  nah i dont think he goes to usf he lives aroun...

Tokenization and stopword removal
python


In [38]:
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
import nltk

In [39]:
# Tokenization and stopword removal
stop_words = set(stopwords.words('english'))

def tokenize_and_remove_stopwords(text):
    tokens = word_tokenize(text)
    filtered_tokens = [word for word in tokens if word not in stop_words]
    return filtered_tokens

df['tokens'] = df['Message'].apply(tokenize_and_remove_stopwords)


Step3:  Transform the processed text data into numerical vectors using CountVectorizer
python

In [40]:
df.columns

Index(['Category', 'Message', 'tokens'], dtype='object')

Step 4: Apply lemmatization

In [41]:
from nltk.stem import PorterStemmer, WordNetLemmatizer
nltk.download('wordnet')


[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [42]:
# Initialize stemmer and lemmatizer
stemmer = PorterStemmer()
lemmatizer = WordNetLemmatizer()


In [43]:
# Apply stemming and lemmatization
stemmer = PorterStemmer()
lemmatizer = WordNetLemmatizer()

def stem_and_lemmatize(tokens):
    stemmed = [stemmer.stem(word) for word in tokens]
    lemmatized = [lemmatizer.lemmatize(word) for word in stemmed]
    return lemmatized


In [44]:
# Apply stemming and lemmatization
df['processed_tokens'] = df['tokens'].apply(stem_and_lemmatize)
df['processed_text'] = df['processed_tokens'].apply(lambda x: ' '.join(x))


In [45]:
df.head()


Category                                            Message  \
0      ham  go until jurong point crazy available only in ...   
1      ham                            ok lar joking wif u oni   
2     spam  free entry in 2 a wkly comp to win fa cup fina...   
3      ham        u dun say so early hor u c already then say   
4      ham  nah i dont think he goes to usf he lives aroun...   

                                              tokens  \
0  [go, jurong, point, crazy, available, bugis, n...   
1                     [ok, lar, joking, wif, u, oni]   
2  [free, entry, 2, wkly, comp, win, fa, cup, fin...   
3      [u, dun, say, early, hor, u, c, already, say]   
4  [nah, dont, think, goes, usf, lives, around, t...   

                                    processed_tokens  \
0  [go, jurong, point, crazi, avail, bugi, n, gre...   
1                       [ok, lar, joke, wif, u, oni]   
2  [free, entri, 2, wkli, comp, win, fa, cup, fin...   
3      [u, dun, say, earli, hor, u, c, alreadi, say]   
4  [nah, dont, think, goe, usf, live, around, tho...   

                                      processed_text  
0  go jurong point crazi avail bugi n great world...  
1                              ok lar joke wif u oni  
2  free entri 2 wkli comp win fa cup final tkt 21...  
3                u dun say earli hor u c alreadi say  
4          nah dont think goe usf live around though

In [46]:
# Create N-gram models
vectorizer_unigram = CountVectorizer(ngram_range=(1, 1))
vectorizer_bigram = CountVectorizer(ngram_range=(1, 2))
vectorizer_trigram = CountVectorizer(ngram_range=(1, 3))


In [47]:
# Transform text data
X_unigram = vectorizer_unigram.fit_transform(df['processed_text'])
X_bigram = vectorizer_bigram.fit_transform(df['processed_text'])
X_trigram = vectorizer_trigram.fit_transform(df['processed_text'])


In [48]:
# Encode the labels
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(df['Category'])

Step 5: Create N-gram models (unigrams, bigrams, trigrams) and transform the text data

In [49]:
# Create n-gram models
vectorizer_unigram = CountVectorizer(ngram_range=(1, 1))
vectorizer_bigram = CountVectorizer(ngram_range=(1, 2))
vectorizer_trigram = CountVectorizer(ngram_range=(1, 3))


In [50]:
# Transform text data
X_unigram = vectorizer_unigram.fit_transform(df['processed_text'])
X_bigram = vectorizer_bigram.fit_transform(df['processed_text'])
X_trigram = vectorizer_trigram.fit_transform(df['processed_text'])


In [51]:
df.head()

Category                                            Message  \
0      ham  go until jurong point crazy available only in ...   
1      ham                            ok lar joking wif u oni   
2     spam  free entry in 2 a wkly comp to win fa cup fina...   
3      ham        u dun say so early hor u c already then say   
4      ham  nah i dont think he goes to usf he lives aroun...   

                                              tokens  \
0  [go, jurong, point, crazy, available, bugis, n...   
1                     [ok, lar, joking, wif, u, oni]   
2  [free, entry, 2, wkly, comp, win, fa, cup, fin...   
3      [u, dun, say, early, hor, u, c, already, say]   
4  [nah, dont, think, goes, usf, lives, around, t...   

                                    processed_tokens  \
0  [go, jurong, point, crazi, avail, bugi, n, gre...   
1                       [ok, lar, joke, wif, u, oni]   
2  [free, entri, 2, wkli, comp, win, fa, cup, fin...   
3      [u, dun, say, earli, hor, u, c, alreadi, say]   
4  [nah, dont, think, goe, usf, live, around, tho...   

                                      processed_text  
0  go jurong point crazi avail bugi n great world...  
1                              ok lar joke wif u oni  
2  free entri 2 wkli comp win fa cup final tkt 21...  
3                u dun say earli hor u c alreadi say  
4          nah dont think goe usf live around though

Step 6: Split the data into training and testing sets

In [52]:
# Split the dataset into training and testing sets
X_train_uni, X_test_uni, y_train_uni, y_test_uni = train_test_split(X_unigram, y, test_size=0.2, random_state=42)
X_train_bi, X_test_bi, y_train_bi, y_test_bi = train_test_split(X_bigram, y, test_size=0.2, random_state=42)
X_train_tri, X_test_tri, y_train_tri, y_test_tri = train_test_split(X_trigram, y, test_size=0.2, random_state=42)

Step 7: Train a classifier on the training set
Using Multinomial Naive Bayes as the classifier

In [57]:
# Initialize the model
model = MultinomialNB()

In [58]:
# Train the model on unigrams
model.fit(X_train_uni, y_train_uni)

MultinomialNB()

In [59]:
# Train the model on bigrams
model_bi = MultinomialNB()
model_bi.fit(X_train_bi, y_train_bi)

MultinomialNB()

In [60]:
# Train the model on trigrams
model_tri = MultinomialNB()
model_tri.fit(X_train_tri, y_train_tri)

MultinomialNB()

Step 8: Evaluate the model on the testing set

In [61]:
# Evaluate the model on unigrams
y_pred_uni = model.predict(X_test_uni)
accuracy_uni = accuracy_score(y_test_uni, y_pred_uni)
precision_uni = precision_score(y_test_uni, y_pred_uni)
recall_uni = recall_score(y_test_uni, y_pred_uni)
f1_uni = f1_score(y_test_uni, y_pred_uni)
report_uni = classification_report(y_test_uni, y_pred_uni)
conf_matrix_uni = confusion_matrix(y_test_uni, y_pred_uni)

print("Unigram Model Performance:")
print(f"Accuracy: {accuracy_uni}")
print(f"Precision: {precision_uni}")
print(f"Recall: {recall_uni}")
print(f"F1-score: {f1_uni}")
print(f"Classification Report:\n {report_uni}")
print(f"Confusion Matrix:\n {conf_matrix_uni}")

Unigram Model Performance:
Accuracy: 0.97847533632287
Precision: 0.9032258064516129
Recall: 0.9395973154362416
F1-score: 0.9210526315789472
Classification Report:
               precision    recall  f1-score   support

           0       0.99      0.98      0.99       966
           1       0.90      0.94      0.92       149

    accuracy                           0.98      1115
   macro avg       0.95      0.96      0.95      1115
weighted avg       0.98      0.98      0.98      1115

Confusion Matrix:
 [[951  15]
 [  9 140]]


In [62]:
# Evaluate the model on bigrams
y_pred_bi = model_bi.predict(X_test_bi)
accuracy_bi = accuracy_score(y_test_bi, y_pred_bi)
precision_bi = precision_score(y_test_bi, y_pred_bi)
recall_bi = recall_score(y_test_bi, y_pred_bi)
f1_bi = f1_score(y_test_bi, y_pred_bi)
report_bi = classification_report(y_test_bi, y_pred_bi)
conf_matrix_bi = confusion_matrix(y_test_bi, y_pred_bi)

print("\nBigram Model Performance:")
print(f"Accuracy: {accuracy_bi}")
print(f"Precision: {precision_bi}")
print(f"Recall: {recall_bi}")
print(f"F1-score: {f1_bi}")
print(f"Classification Report:\n {report_bi}")
print(f"Confusion Matrix:\n {conf_matrix_bi}")


Bigram Model Performance:
Accuracy: 0.9721973094170404
Precision: 0.8597560975609756
Recall: 0.9463087248322147
F1-score: 0.9009584664536741
Classification Report:
               precision    recall  f1-score   support

           0       0.99      0.98      0.98       966
           1       0.86      0.95      0.90       149

    accuracy                           0.97      1115
   macro avg       0.93      0.96      0.94      1115
weighted avg       0.97      0.97      0.97      1115

Confusion Matrix:
 [[943  23]
 [  8 141]]


In [63]:
# Evaluate the model on trigrams
y_pred_tri = model_tri.predict(X_test_tri)
accuracy_tri = accuracy_score(y_test_tri, y_pred_tri)
precision_tri = precision_score(y_test_tri, y_pred_tri)
recall_tri = recall_score(y_test_tri, y_pred_tri)
f1_tri = f1_score(y_test_tri, y_pred_tri)
report_tri = classification_report(y_test_tri, y_pred_tri)
conf_matrix_tri = confusion_matrix(y_test_tri, y_pred_tri)

print("\nTrigram Model Performance:")
print(f"Accuracy: {accuracy_tri}")
print(f"Precision: {precision_tri}")
print(f"Recall: {recall_tri}")
print(f"F1-score: {f1_tri}")
print(f"Classification Report:\n {report_tri}")
print(f"Confusion Matrix:\n {conf_matrix_tri}")


Trigram Model Performance:
Accuracy: 0.9632286995515695
Precision: 0.8068181818181818
Recall: 0.9530201342281879
F1-score: 0.8738461538461537
Classification Report:
               precision    recall  f1-score   support

           0       0.99      0.96      0.98       966
           1       0.81      0.95      0.87       149

    accuracy                           0.96      1115
   macro avg       0.90      0.96      0.93      1115
weighted avg       0.97      0.96      0.96      1115

Confusion Matrix:
 [[932  34]
 [  7 142]]


Deliverables Summary
Preprocessed Dataset: The dataset with preprocessed text (normalization, tokenization, stopword removal, stemming, and lemmatization).

Code and Documentation: Provided in the above cells.

#Report Summary:

###Unigram Model Performance:
Accuracy: [0.97847533632287]

Precision: [0.9032258064516129]

Recall: [0.9395973154362416]

F1-score: [0.9210526315789472]

###Bigram Model Performance:
Accuracy: [0.9721973094170404]

Precision: [0.8597560975609756]

Recall: [0.9463087248322147]

F1-score: [0.9009584664536741]

###Trigram Model Performance:
Accuracy: [0.9632286995515695]

Precision: [0.8068181818181818]

Recall: [0.9530201342281879]

F1-score: [0.8738461538461537]










#Conclusion:
The Unigram model outperforms both the Bigram and Trigram models in terms of accuracy, precision, and F1-score, making it the best model for this spam detection task. While the Trigram model has the highest recall, the Unigram model's overall performance, as indicated by its higher F1-score, suggests that it balances precision and recall more effectively. Therefore, the Unigram model is the best choice based on the given results.